src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -398,14 +471,377 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
             }
         }
 
-        (self.qualif, Lrc::new(promoted_temps))
+        let qualif = (self.mut_interior_checker.mut_interior,
+                      self.drop_checker.needs_drop,
+                      self.not_const);
+        (qualif, Lrc::new(promoted_temps))
+    }
+}
+
+/// Accumulates an Rvalue or a Call's effects into `self.mut_interior`.
+impl<'a, 'tcx> Visitor<'tcx> for MutInteriorChecker<'a, 'tcx, 'tcx> {
+    fn visit_local(&mut self,
+                   &local: &Local,
+                   _: PlaceContext<'tcx>,
+                   _: Location) {
+        let kind = self.mir.local_kind(local);
+        match kind {
+            LocalKind::ReturnPointer => {}
```

Comment:
@oli-obk @eddyb Should the arm for this actually be something like `self.mut_interior |= self.local_mut_interior.contains(&RETURN_PLACE);`?